draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -1172,86 +1177,141 @@ Steps:
 
 ### Elligator2 Method {#elligator2}
 
-The map2curve\_elligator2(alpha) implements the Elligator2 {{BHKL13}} for
+The function map2curve\_elligator2(alpha) implements the Elligator2 {{BHKL13}} for
 curves defined by y^2 = x^3 + A * x^2 + B * x such that A * B * (A^2 - 4 * B) != 0.
 In particular, this method applies to the Montgomery curves y^2 = x^3 + A * x^2 + x
 setting B=1.
 
-Preconditions: A Montgomery curve such that A!=0.
+Preconditions: A Montgomery curve such that A != 0 and A^2 - 4 is non-square in F.
 
 Input: alpha, an octet string to be hashed.
 
-Constants: A and B, the parameters of the curve; N, a non-square in F.
+Constants:
+
+- A and B, the parameters of the curve
+
+- Z the smallest (in absolute value) non-square in F, breaking ties by choosing
+  the positive value.
 
 Output: (x, y), a point on E.
 
+Exceptions: The exceptional case is Z * u^2 == -1, i.e., 1 + Z * u^2 == 0.
+Implementations must detect this case and set x1 = -A.
+Note that this can only happen when q = 3 (mod 4).
+
 Operations:
 
 ~~~
 1.   u = hash2base(alpha)
-2.  x1 = -A / (1 + N * u^2)
-3. gx1 = x1^3 + A * x1^2 + B * x1
-4.  x2 = -x1 - A
-5. gx2 = x2^3 + A * x2^2 + B * x2
-6.   e = gx1^((q - 1) / 2)
-7. If is_square(gx1), set x = x1 and y = -e * sqrt(gx1)
-8. If is_square(gx2), set x = x2 and y = -e * sqrt(gx2)
-9. Output h * (x, y)
+2.  x1 = -A * inv0(1 + Z * u^2)
+3. If x1 == 0, set x1 = -A.
+4. gx1 = x1^3 + A * x1^2 + B * x1
+5.  x2 = -x1 - A
+6. gx2 = x2^3 + A * x2^2 + B * x2
+7.   e = gx1^((q - 1) / 2)
+8. If is_square(gx1), set x = x1 and y = -e * sqrt(gx1)
+9. If is_square(gx2), set x = x2 and y = -e * sqrt(gx2)
+10. Output h * (x, y)
 ~~~
 
-#### Implementation
+#### Implementation, q=3 (mod 4)
```

Comment:
nit: missing spaces around `=`

Reply:
this seems to be a pervasive issue that I didn't quite manage to fix. I opened #105 about this, because fixing all of them would be a pretty huge diff to tack on to this one.